EZP-29751: Wrong first URL generation in exotic languages

glye opened the pull request:
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-29751](https://jira.ez.no/browse/EZP-29751)
| **Bug/Improvement**| yes
| **New feature**    | no
| **Target version** | `6.x`/`7.x`
| **BC breaks**      | no
| **Tests pass**     | yes
| **Doc needed**     | no

With languages like Japanese and Chinese, names typically get converted to empty string by the URL alias handler. For the first such name in any location, the empty string is accepted as a unique string, and saved as is, making this content unaccessible by URL alias. The second, third etc names in this location will get aliases `2`, `3`, etc. because of code that avoids duplicate aliases.

Solution: Ensure that empty string is not accepted as valid URL alias. So the first such name gets the alias `1` (the number is taken from the deduplication counter).

glye commented:
Squashed commits to hopefully get stuck CS test going again, no code change. Unit tests were passing before.

alongosz commented on eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Handler.php:
```diff
@@ -187,7 +187,7 @@ private function internalPublishUrlAliasForLocation(
             $newText = '';
             if ($locationId != self::CONTENT_REPOSITORY_ROOT_LOCATION_ID) {
-                $newText = $name . ($uniqueCounter > 1 ? $uniqueCounter : '');
+                $newText = $name . ($name == '' || $uniqueCounter > 1 ? $uniqueCounter : '');
```
How would this behave in case there was more than one names which results in an empty string? My assumption is that `$uniqueCounter` still would be 1, which worries me a bit.

glye replied:
@alongosz Duplicates can happen if there are pre-existing db entries. The counter will be increased by 1 then, see where it deals with UniqueConstraintViolationException 30 lines below, and at the end of the while loop.

alongosz commented on eZ/Publish/Core/Persistence/Legacy/Tests/Content/UrlAlias/UrlAliasHandlerTest.php:
```diff
@@ -1170,6 +1170,76 @@ public function testPublishUrlAliasCreatesUniqueAlias()
+     * @depends testPublishUrlAliasForLocation
+     * @group publish
+     */
+    public function testPublishUrlAliasCreatesNonEmptyAlias()
```
`eZ/Publish/Core/Persistence/Legacy/Tests` namespace is not a good place for tests which operate on database. It means that it will be tested on sqlite only. Such tests are integration tests and should be done in e.g. `\eZ\Publish\API\Repository\Tests\URLAliasServiceTest`. Alternatively you could change it into unit test with mocks.

_Yes, I know you've copy-pasted this from other tests like this, but this didn't change the fact it was oversight by someone who started writing these tests. Sorry :(_

glye replied:
@alongosz Thanks (no really, I mean it - this ought to be properly tested in all dbs :) )

glye replied:
Removed test, added test in `\eZ\Publish\API\Repository\Tests\URLAliasServiceTest`, verified locally that it fails before the test and passes after.

glye commented:
Update: Removed sqlite test, added test in `\eZ\Publish\API\Repository\Tests\URLAliasServiceTest`, verified locally that it fails before the patch and passes after. @alongosz

alongosz commented on eZ/Publish/API/Repository/Tests/URLAliasServiceTest.php:
```diff
@@ -1215,6 +1215,55 @@ public function testRefreshSystemUrlAliasesForContentsWithUpdatedContentTypes()
+    public function testCreateNonLatinNonEmptyUniqueAliases()
+    {
+        $urlAliasService = $this->getRepository()->getURLAliasService();
+        $locationService = $this->getRepository()->getLocationService();
```
```php
        $repository = $this->getRepository();
        $urlAliasService = $repository->getURLAliasService();
        $locationService = $repository->getLocationService();
```
Otherwise you're initializing database twice :upside_down_face:

glye commented:
Fixed double database initialisation. @alongosz

glye commented:
@micszo If none of them had any a-z, 0-9 characters in their names, and none of them were at the same location, then that is correct, yes.

micszo commented:
Yes, they were all Arabic or all Japanese and at different levels of a subtree. So seems it's ok.

micszo commented:
@mateuszbieniek have you tried this on v2.2.3? the results are different than on v1.7.8 and v2.3.2.

micszo commented:
As in the _text_ field value in _ezurlalias_ml_ gets filed but the URL tab in UI doesn't display any System URLs.

glye commented:
Huh. Can it be a caching issue?

micszo commented:
Negative. Removing and reloading cache doesn't change it. (Private mode neither).

With patch after republishing the value is added in db but nothing is displayed in UI.
![screenshot 2018-12-05 at 10 23 49](https://user-images.githubusercontent.com/13622502/49503819-971fae80-f878-11e8-9ee9-5faac5dfcd45.png)
![screenshot 2018-12-05 at 10 23 16](https://user-images.githubusercontent.com/13622502/49503830-9e46bc80-f878-11e8-8ca1-dd1bcf12101a.png)
![screenshot 2018-12-05 at 10 23 22](https://user-images.githubusercontent.com/13622502/49503844-a999e800-f878-11e8-9e34-40056e636716.png)

Interesting thing is that with or without the patch the "first level" directories have the value in db but also nothing is displayed in UI.
![screenshot 2018-12-05 at 10 32 31](https://user-images.githubusercontent.com/13622502/49504036-2f1d9800-f879-11e8-9d1c-ef4e4afc49ea.png)
I'm guessing it's a v2.2 admin-ui specific problem.

mateuszbieniek commented:
@micszo did you mean to reproduce the issue with 2.2 or try the fix on 2.2? In both cases: I did not :wink:

micszo commented:
yes, both, I tried both. 🙂

micszo commented:
basically on 1.7, 1.13 and 2.3 everything works find in db and UI but on 2.2 in db it's fine but something is off in UI.

glye commented:
It if doesn't actually make anything *worse* in 2.2, I'd leave this issue as a follow-up, then.

micszo commented:
I was leaning in the same direction. 🙂 It would probably in a different bundle.

glye commented:
True. 2.2 is past end-of-maintenance, though end-of-life is not until end of March. This one may not be important enough for a past-maintenance fix. But registering it as a known issue is good, I guess.

micszo commented:
It's already linked with this issue in JIRA.
